pkg/products/threescale/reconciler.go

```diff
@@ -2411,3 +2022,203 @@ func (r *Reconciler) ensureDeploymentConfigsReady(ctx context.Context, serverCli
+	if err != nil {
+		r.log.Errorf("Failed to create envoyconfig for apicast", l.Fields{"APICast": ApicastClusterName}, err)
+		return integreatlyv1alpha1.PhaseFailed, err
+	}
+
+	// backend-listener cluster
+	backendClusterResource := ratelimit.CreateClusterResource(
+		BackendContainerAddress,
+		BackendClusterName,
+		BackendContainerPort,
+	)
+
+	backendHTTPFilters, _ := getBackendListenerHTTPFilters()
+	// backend listener listener
+	backendFilters, _ := getListenerResourceFilters(
+		getBackendListenerVitualHosts(BackendClusterName),
+		backendHTTPFilters,
+	)
+	backendListenerResource := ratelimit.CreateListenerResource(
+		BackendListenerName,
+		BackendEnvoyProxyAddress,
+		BackendEnvoyProxyPort,
+		backendFilters,
+	)
+
+	// create envoy config for apicast
+	backendProxyConfig := ratelimit.NewEnvoyConfig(BackendClusterName, r.Config.GetNamespace(), BackendNodeID)
+	err = backendProxyConfig.CreateEnvoyConfig(ctx, serverClient, []*envoyapi.Cluster{backendClusterResource, ratelimitClusterResource}, []*envoyapi.Listener{backendListenerResource}, installation)
+	if err != nil {
+		r.log.Errorf("Failed to create envoyconfig for backend-listener", l.Fields{"BackendListener": BackendClusterName}, err)
+		return integreatlyv1alpha1.PhaseFailed, err
+	}
+
+	return integreatlyv1alpha1.PhaseCompleted, nil
+}
+
+func (r *Reconciler) getRateLimitServiceCR(ctx context.Context, serverClient k8sclient.Client) (*corev1.Service, error) {
+	rateLimitService := &corev1.Service{}
+	marin3rConfig, err := r.ConfigManager.ReadMarin3r()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load marin3r config in 3scale reconciler: %v", err)
+	}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{
+		Namespace: marin3rConfig.GetNamespace(),
+		Name:      "ratelimit",
+	}, rateLimitService)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to rate limiting service: %v", err)
+	}
+
+	return rateLimitService, nil
+}
+
+func getRatelimitServicePort(rateLimitService *corev1.Service) int {
+	for _, port := range rateLimitService.Spec.Ports {
+		if port.Name == "grpc" {
+			return port.TargetPort.IntValue()
+		}
+	}
+	return 0
+}
+
+func (r *Reconciler) createBackendListenerProxyService(ctx context.Context, serverClient k8sclient.Client) error {
+	serviceName := "backend-listener-proxy"
+	backendListenerService := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, serverClient, backendListenerService, func() error {
+		owner.AddIntegreatlyOwnerAnnotations(backendListenerService, r.installation)
+		backendListenerService.Spec.Ports = []corev1.ServicePort{
+			{
+				Name:       "http",
+				Protocol:   "TCP",
+				Port:       3001,
+				TargetPort: intstr.FromInt(3001),
```

Comment:
those can also be defined as constant up top, probably :thinking:

Reply:
Good catch